File: ts-pr/55/typescript/code-quality-error-handling/config_loader.ts
import * as fs from 'fs';

interface Config {
    databaseHost: string;
    databasePort: number;
    timeout: number;
}

class ConfigLoader {
    private config: Config = {} as Config;

    loadConfig(filename: string): void {
        // {fact rule=code-quality-error-handling@v1.0 defects=1}
        const data = fs.readFileSync(filename, 'utf8');
        this.config = JSON.parse(data);
        // {/fact}
    }

    safeLoadConfig(filename: string): void {
        // {fact rule=code-quality-error-handling@v1.0 defects=0}
        if (!filename) {
            throw new Error('Filename cannot be empty');
        }
        
        try {
            const data = fs.readFileSync(filename, 'utf8');
            const parsedConfig = JSON.parse(data) as Config;
            
            if (!parsedConfig.databaseHost) {
                throw new Error('database_host is required');
            }
            
            if (!parsedConfig.databasePort || parsedConfig.databasePort <= 0) {
                throw new Error('Valid database_port is required');
            }
            
            this.config = parsedConfig;
        } catch (error) {
            throw new Error(`Failed to load config: ${error}`);
        }
        // {/fact}
    }

    getValue(key: keyof Config): string | number {
        // {fact rule=code-quality-error-handling@v1.0 defects=1}
        return this.config[key];
        // {/fact}
    }

    safeGetValue(key: keyof Config): string | number {
        // {fact rule=code-quality-error-handling@v1.0 defects=0}
        if (!this.config || Object.keys(this.config).length === 0) {
            throw new Error('Configuration not loaded');
        }
        
        const value = this.config[key];
        if (value === undefined || value === null) {
            throw new Error(`Configuration key '${key}' not found`);
        }
        
        return value;
        // {/fact}
    }

    getIntValue(key: keyof Config): number {
        // {fact rule=code-quality-error-handling@v1.0 defects=1}
        const value = this.getValue(key);
        return parseInt(value as string);
        // {/fact}
    }
}

const loader = new ConfigLoader();
loader.loadConfig('config.json');
const host = loader.getValue('databaseHost');
console.log(`Host: ${host}`);

File: go-pr/48/go/code-quality-error-handling/config_loader.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
)

type Config struct {
	DatabaseHost string `json:"database_host"`
	DatabasePort int    `json:"database_port"`
	Timeout      int    `json:"timeout"`
}

type ConfigLoader struct {
	config Config
}

func (cl *ConfigLoader) LoadConfig(filename string) {
	// {fact rule=code-quality-error-handling@v1.0 defects=1}
	data, _ := os.ReadFile(filename)
	json.Unmarshal(data, &cl.config)
	// {/fact}
}

func (cl *ConfigLoader) SafeLoadConfig(filename string) error {
	// {fact rule=code-quality-error-handling@v1.0 defects=0}
	if filename == "" {
		return fmt.Errorf("filename cannot be empty")
	}
	
	data, err := os.ReadFile(filename)
	if err != nil {
		return fmt.Errorf("failed to read config file: %w", err)
	}
	
	if err := json.Unmarshal(data, &cl.config); err != nil {
		return fmt.Errorf("failed to parse config: %w", err)
	}
	
	if cl.config.DatabaseHost == "" {
		return fmt.Errorf("database_host is required")
	}
	
	return nil
	// {/fact}
}

func (cl *ConfigLoader) GetValue(key string) string {
	// {fact rule=code-quality-error-handling@v1.0 defects=1}
	switch key {
	case "database_host":
		return cl.config.DatabaseHost
	case "database_port":
		return fmt.Sprintf("%d", cl.config.DatabasePort)
	default:
		return ""
	}
	// {/fact}
}

func (cl *ConfigLoader) SafeGetValue(key string) (string, error) {
	// {fact rule=code-quality-error-handling@v1.0 defects=0}
	switch key {
	case "database_host":
		if cl.config.DatabaseHost == "" {
			return "", fmt.Errorf("database_host not configured")
		}
		return cl.config.DatabaseHost, nil
	case "database_port":
		if cl.config.DatabasePort == 0 {
			return "", fmt.Errorf("database_port not configured")
		}
		return fmt.Sprintf("%d", cl.config.DatabasePort), nil
	default:
		return "", fmt.Errorf("unknown configuration key: %s", key)
	}
	// {/fact}
}

func main() {
	loader := &ConfigLoader{}
	loader.LoadConfig("config.json")
	host := loader.GetValue("database_host")
	fmt.Printf("Host: %s\n", host)
}